esmvaltool/utils/recipe_filler.py

```diff
@@ -0,0 +1,684 @@
+                                            end_year)
+
+    if not valid_files:
+        logger.warning("No data found to fully cover start "
+                       f"{start_year} / end {end_year} as requested!")
+
+    return valid_files
+
+
+def _resolve_latestversion(dirname_template):
+    """Resolve the 'latestversion' tag."""
+    if '{latestversion}' not in dirname_template:
+        return dirname_template
+
+    # Find latest version
+    part1, part2 = dirname_template.split('{latestversion}')
+    part2 = part2.lstrip(os.sep)
+    part1_contents = glob(part1)
+    if part1_contents:
+        versions = os.listdir(part1_contents[0])
+        versions.sort(reverse=True)
+        for version in ['latest'] + versions:
+            dirname = os.path.join(part1, version, part2)
+            if glob(dirname):
+                return dirname
+
+    return dirname_template
+
+
+def list_all_files(file_dict, cmip_era):
+    """
+    List all files that match the dataset dictionary.
+
+    Function that returnes all files that are determined by a
+    file_dict dictionary; file_dict is keyed on usual parameters
+    like `dataset`, `project`, `mip` etc; glob.glob is used
+    to find files; speedup is achieved by replacing wildcards
+    with values from CMOR tables.
+
+    Parameters
+    ----------
+    file_dict: dict
+        Dictionary to hold dataset specifications.
+
+    cmip_era: str
+        Either CMIP5 or CMIP6.
+
+    Returns
+    -------
+    list:
+        List of found files.
+
+    """
+    mip = file_dict['mip']
+    short_name = file_dict['short_name']
+    try:
+        frequency = CMOR_TABLES[cmip_era].get_variable(mip,
+                                                       short_name).frequency
+        realms = CMOR_TABLES[cmip_era].get_variable(mip,
+                                                    short_name).modeling_realm
+    except AttributeError:
+        logger.warning(f"Could not find {cmip_era} CMOR table "
+                       f"for variable {short_name} with mip {mip}")
+        return []
+    file_dict['frequency'] = frequency
+
+    basepaths = _determine_basepath(cmip_era)
+    all_files = []
+
+    for basepath in basepaths:
+        new_path = basepath[:]
+
+        # could have multiple realms
+        for realm in realms:
+            file_dict['modeling_realm'] = realm
+
+            # load all the files in the custom dict
+            for key, value in file_dict.items():
+                new_path = new_path.replace('{' + key + '}', str(value))
+            new_path = _resolve_latestversion(new_path)
+            if new_path.startswith("~"):
+                new_path = os.path.expanduser(new_path)
+                if not new_path.startswith(os.sep):
+                    logger.error("Could not expand ~ to user home dir "
+                                 "please expand it in the config user file!")
+                    sys.exit(1)
```

Comment:
```suggestion
                    raise ValueError("Could not expand ~ to user home dir please expand it in the config user file!")
```

Reply:
ok probably raising is more useful, I was thinking sysexiting is better for command-line tools that may be run from automated scripts, but actually a `sys.exit(X)` is a raiser of an exception too - so no benefits